File: 2020/Team/БГУИР-2020-semifinal/A/a.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using ld = long double;
using ull = unsigned long long;
#define all(x) (x).begin(),(x).end()
#define int ll
const int MAXN = 21;

ll fact[MAXN];

void build() {
	fact[0] = 1;
	for (int i = 1; i < MAXN; i++)
		fact[i] = i * fact[i - 1];
}

int n, a, b;
int p[MAXN];
bool used[MAXN];

int calc_cycles() {
	int ans = 0;
	for (int i = 0; i < n; i++) {
		if (used[i]) continue;
		int v = i;
		ans++;
		while (!used[v]) {
			used[v] = true;
			v = p[v] - 1;
		}
	}
	return ans;
}

int cost_cycle(int cnt) {
	return n - cnt;
}

ll get(ll n, ll k) {
	if (n == 0 && k == 0) return 1;
	if (n == 0 || k == 0) return 0;
	return get(n - 1, k - 1) + (n - 1) * get(n - 1, k);
}

ld get_path(int cnt) {
	ld ans = (ld) fact[n] / get(n, cnt);
	return ans;
}

signed main() {
	ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);
	cout.precision(20);
	cout.setf(ios::fixed);
	build();
	cin >> n >> a >> b;
	for (int i = 0; i < n; i++)
		cin >> p[i];
	ld ans = cost_cycle(calc_cycles()) * a;
	//cout << ans << endl;
	for (int i = 1; i <= n; i++) {
		//cout << i << " " << cost_cycle(i) << endl;
		//cout << (ld) get_path(i) * b << endl;
		ans = min(ans, (ld)get_path(i) * b + cost_cycle(i) * a);
	}
	cout << ans << endl;
	return 0;
}

File: 2020/Team/БГУИР-2020-semifinal/A/a.py
from decimal import *

MAXN = 21

fact = [0 for i in range(MAXN)]

def build():
    fact[0] = 1
    for i in range(1, MAXN):
        fact[i] = fact[i - 1] * i

n, a, b = None, None, None
p = [0 for i in range(MAXN)]
used = p.copy()

def calc_cycles():
    ans = 0
    for i in range(n):
        if used[i]:
            continue
        v = i
        ans += 1
        while not used[v]:
            used[v] = 1
            v = p[v] - 1
    return ans

def cost_cycle(cnt):
    return n - cnt

def get(n, k):
    if n == 0 and k == 0:
        return 1
    if n == 0 or k == 0:
        return 0
    return get(n - 1, k - 1) - (n - 1) * get(n - 1, k)

def fget(n, k):
    return abs(get(n, k))

def get_path(cnt):
    ans = Decimal(fact[n]) / fget(n, cnt)
    return ans

def main():
    global n, a, b, p
    build()
    n, a, b = input().split()
    n = int(n)
    a, b = map(Decimal, [a, b])
    p = list(map(int, input().split()))
    ans = cost_cycle(calc_cycles()) * a
    fans = Decimal(0)
    for i in range(1, n + 1):
        if (i != calc_cycles()):
            fans += get_path(i) * b + cost_cycle(i) * a
    ans = min(ans, fans / n)
    print(ans)

main()
